File: src/Monet/MOSound.m
#import <Monet/MOSound.h>

#import <SDL_mixer.h>

struct MOSoundData
{
	Mix_Chunk *chunk;
};

@implementation MOSound

- (id)initWithFilename: (NSString *)aName
{
	if ((self = [super init]))
	{
		soundData = calloc(1, sizeof (struct MOSoundData));

		soundData->chunk = Mix_LoadWAV([aName UTF8String]);
	}

	return self;
}

- (void)dealloc
{
	Mix_FreeChunk(soundData->chunk);

	[super dealloc];
}

#pragma mark -

+ (MOSound *)named: (NSString *)aName
{
	return [[[self alloc] initWithFilename: aName] autorelease];
}

#pragma mark -

- (void)play
{
	int channel = Mix_PlayChannel(-1, soundData->chunk, 0);
}

@end

File: src/Monet/MOSound.c
#include <Monet/MOSound.h>

#include <Monet/Common.h>

struct _MOSound
{
	COGuts    guts;

	Mix_Chunk *chunk;
};

void _MOSoundDestroy(void *aSound);

COClass MOSoundClass = {
	.superclass = NULL,
	.destructor = &_MOSoundDestroy
};

MOSound *MOSoundCreateFromFilename(char *aFilename)
{
	MOSound *sound = malloc(sizeof (MOSound));
	COInit(sound, &MOSoundClass);

	sound->chunk = Mix_LoadWAV(aFilename);

	return sound;
}

void _MOSoundDestroy(void *aSound)
{
	MOSound *sound = (MOSound *)aSound;

	Mix_FreeChunk(sound->chunk);
}

void MOSoundPlay(MOSound *aSound)
{
	Mix_PlayChannel(-1, aSound->chunk, 0);
}
